Add a config to choose which components to extract strings from.

pcorpet opened the pull request:
For #33.

gilbsgilbs commented on docs/templates/configuration.md.yml:
````diff
@@ -75,6 +75,22 @@ configOptions:
           ```
+  - name: transComponentNames
+    type: "{sourceModule: string; importName: string}[]"
+    description: Possible import names of `Trans` components. Note that, for now, it does a pretty dumb job at resolving the source module so you should list all ways you have to import your components.
````
I think this is too much of a limitation, even for a MVP. In particular, having `import {CustomTrans} from "./customTransComponent"` work when `import {CustomTrans} from "../customTransComponent"` doesn't would be very confusing in my opinion. It should resolve relative modules and I'd say it should be relative to the `.babelrc` or the `package.json`.

pcorpet replied:
OK, I've made it work for relative imports. However:
- I'm not sure to which file it's related (I looked at outputPath's doc that does not say much either)
- it won't work if callers use weird ways of getting their modules (e.g. `../components/i18n` while being in the components folder) as it's only trying the most direct relative path. However going further would require to rewrite the referencesImport function

gilbsgilbs commented on docs/templates/configuration.md.yml:
```diff
@@ -75,6 +75,22 @@ configOptions:
+  - name: transComponentNames
+    type: "{sourceModule: string; importName: string}[]"
+    description: Possible import names of `Trans` components. Note that, for now, it does a pretty dumb job at resolving the source module so you should list all ways you have to import your components.
+    defaultValue: "[{sourceModule: 'react-i18next', importName: 'Trans'}]"
```
Did you choose this API for forward-compatibility purpose? I feel like the following API would be more comfortable to use:
```javascript
[
  ['react-i18next', 'Trans'],
  ['./src/components/CustomTransComponent', 'Trans'],
]
```
For forward-compatibility, a third value could be added to the array if it's needed someday. WDYT?

pcorpet replied:
Nope, I did that mostly to be as explicit as possible. I've switched to your way.

gilbsgilbs commented on src/config.ts:
```diff
@@ -77,5 +81,11 @@ export function parseConfig(opts: Partial<Config>): Config {
+    transComponentNames: coalesce(opts.transComponentNames, [
+      {
+        sourceModule: 'react-i18next',
+        importName: 'Trans',
```
I wonder how much we want `Trans` component to be part of the default value. On the one hand it allows the user to disable raw `Trans` component extraction straight away which is good, but on the other hand it forces the user to specify the default value again if they just want to extend it. I'd like to hear your rationale.

pcorpet replied:
My rationale is my usecase, which is: overriding the Trans component completely (because of https://github.com/i18next/react-i18next/issues/1014 if you want to know). So I want to disactivate the basic Trans.

Note that if it's not in the default, you would need another setting to turn it off.

gilbsgilbs commented on src/extractors/transComponent.ts:
```diff
@@ -21,17 +21,18 @@ import { Config } from '../config';
-    openingElement.get('name'),
-    'react-i18next',
-    'Trans',
+  return transComponentNames.some(({ sourceModule, importName }): boolean =>
```
I feel concerned about how the complexity of `Trans` component extraction will evolve if we let the module resolution code here. I mean `TransComponent` extractor is already full of procedural code, recursions, corner-cases and subtilities. An alternative approach I was considering was to have a separate `customTransCompoent` extractor that relies on the `transComponent` extractor but skips the check. That relates to how other `t()` extractors work (like `translationRenderProp` for instance), and it may slightly reduce complexity by improving separation of concerns.

pcorpet replied:
I believe that there are two main features now:
- checking whether a node is to be extracted
- actually extracting

The first one is contained in a single function and we could easily group the second in a single function to be exported if it were to be used in another extractor.

However I don't see which other extractor you have in mind as they would basically be the same, just acting with a different config. I think it's different (for now) in `t` case, as the context in which you extract from t might give you different indications.

Anyway happy to refactor in 2 files if you think it's important. If so please confirm the way you would like to activate one extractor over the other (maybe it's related to your wish of having a default value being an empty array).

gilbsgilbs commented on src/config.ts:
```diff
@@ -21,6 +21,10 @@ export interface Config {
   discardOldKeys: boolean;
   jsonSpace: string | number;
   enableExperimentalIcu: boolean;
+  transComponentNames: readonly {
```
This reminds me that I underuse the `readonly` keyword :+1: .

gilbsgilbs commented:
This PR was merged with rebase in #125.

Note that a changed a few minor things. Most importantly:
- I moved custom Trans components to a separate extractor. I think it will prevent the introduction of technical debt in the long run, and it's closer to the behavior of the `tFunctionNames` in reality. As a side-effect, it won't be possible to disable extraction of raw `Trans` component, but since you don't want to use it in your code base, I don't think it's problematic anyways.
- I changed the option name. I know you tried to make it consistent with `tFunctionNames`, but actually, I don't like the `tFunctionNames` anymore and I think it should be changed at some point (see #124). Like `customTransComponents`, `tFunctionNames` should be a separate extractor and the default extractor should read the imports.

Thank you for this very nice PR @pcorpet :tada:! You did 99% of the job, I just moved things around. It should be released tonight (0.6.0).

pcorpet commented:
> * As a side-effect, it won't be possible to disable extraction of raw `Trans` component, but since you don't want to use it in your code base, I don't think it's problematic anyways.

Actually I did want to exclude the raw Trans component, I guess I'll need to do that with comment hints. Anyway, thanks for finishing the job, and thanks in advance for the release.

gilbsgilbs commented:
> Actually I did want to exclude the raw Trans component

May I ask why?

pcorpet commented:
I have replaced the Trans component by my own, so the places where the original Trans component is used is either in my own trans component code, or is a mistake.

As for why I've replaced the original Trans component, cf https://github.com/i18next/react-i18next/issues/1014

gilbsgilbs commented:
> either in my own trans component code, or is a mistake

Yeah, I think comment hints are fine for Trans component wrappers. For mistakes, I think it's rather the job of a linter (something like [no-restricted-imports](https://eslint.org/docs/rules/no-restricted-imports)).
